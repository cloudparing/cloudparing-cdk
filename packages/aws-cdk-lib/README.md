# Cloudparing AWS CDK Construct Library

CDK constructs focused on monitoring and reducing cloud infrastructure costs.

## Installation

<details><summary><strong>TypeScript</strong></summary>

> https://www.npmjs.com/package/@cloudparing/aws-cdk-lib

In your `package.json`:

```json
{
  "dependencies": {
    "@cloudparing/aws-cdk-lib": "^0.0.0",

    // peer dependencies
    "aws-cdk-lib": "^2.65.0",
    "constructs": "^10.0.5"

    // ...your other dependencies...
  }
}
```
</details>

## Features

### Cost and Usage Report (CUR) v2 Report

```ts

    const curBucket =
      props.bucket ||
      new cdk.aws_s3.Bucket(this, `Cur2Bucket`, {
        bucketName: `cur2-${cdk.Aws.ACCOUNT_ID}-${cdk.Aws.REGION}`,
      });

    new Cur2ExportDefinition(this, `Cur2ExportDefinition`, {
      name: 'cur2-daily-csv',
      description: props.exportDescription,
      bucket: curBucket,
    });

```

### Legacy Cost and Usage Report (CUR)

An L2 construct and CDK stack for configuring the AWS Cost and Usage Report.

**NOTE: The CUR report must be deployed to us-east-1**

**Example using the ReportDefinition construct:**
```
    new ReportDefinition(this, `ReportDefinitionId`, {
      reportName: "CUR Report",
      s3Bucket: yourS3Bucket,
    });
```

**Example using the provided Stacks:**
```
import { CurStack, Cur2Stack } from '@cloudparing/aws-cdk-lib';
import * as cdk from 'aws-cdk-lib';

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new cdk.App();

new CurStack(app, 'curStack-dev', {
  env: devEnv,
  stackName: 'CurStack',
  reportName: 'cur-daily-csv',
});

new Cur2Stack(app, 'cur2Stack-dev', {
  env: devEnv,
  stackName: 'Cur2Stack',
  name: 'cur2-daily-csv',
});

app.synth();

```

Both the L2 construct and the stack provide sensible defaults which can be customized by setting props.

```
    new ReportDefinition(this, `ReportDefinitionId`, {
      reportName: "CUR Report",
      s3Bucket: yourS3Bucket,
      format: 'Parquet'
    });
```


## License

This project is licensed under the Apache-2.0 License.