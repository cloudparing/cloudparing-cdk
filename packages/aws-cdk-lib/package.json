{
  "name": "@cloudparing/aws-cdk-lib",
  "description": "The cloudparing CDK library",
  "repository": {
    "type": "git",
    "url": "github.com/cloudparing/cloudparing-cdk"
  },
  "scripts": {
    "build": "npx projen build",
    "bump": "npx projen bump",
    "compat": "npx projen compat",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "docgen": "npx projen docgen",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "package-all": "npx projen package-all",
    "package:js": "npx projen package:js",
    "package:python": "npx projen package:python",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "test": "npx projen test",
    "test:watch": "npx projen test:watch",
    "unbump": "npx projen unbump",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
    "projen": "npx projen"
  },
  "author": {
    "name": "James Ayvaz",
    "email": "james@cloudparing.net",
    "organization": false
  },
  "devDependencies": {
    "@types/aws-lambda": "^8.10.133",
    "@types/jest": "^29",
    "@types/node": "^18",
    "@typescript-eslint/eslint-plugin": "^6",
    "@typescript-eslint/parser": "^6",
    "aws-cdk-lib": "2.112.0",
    "constructs": "10.3.0",
    "eslint": "^8",
    "eslint-import-resolver-typescript": "^3.5.5",
    "eslint-plugin-import": "^2.27.5",
    "jest": "^29",
    "jest-junit": "^15",
    "jsii": "~5.2.0",
    "jsii-diff": "^1.84.0",
    "jsii-docgen": "^9.0.10",
    "jsii-pacmak": "^1.84.0",
    "jsii-rosetta": "~5.2.0",
    "projen": "^0.71.110",
    "ts-jest": "^29",
    "typescript": "^5.1.3"
  },
  "peerDependencies": {
    "aws-cdk-lib": "^2.112.0",
    "constructs": "^10.3.0"
  },
  "dependencies": {
    "@aws-sdk/client-bcm-data-exports": "^3.502.0",
    "@aws-sdk/client-s3": "^3.504.0",
    "aws-lambda": "^1.0.7"
  },
  "bundledDependencies": [
    "@aws-sdk/client-bcm-data-exports",
    "@aws-sdk/client-s3",
    "aws-lambda"
  ],
  "keywords": [
    "aws",
    "cdk",
    "cost",
    "cur"
  ],
  "main": "lib/index.js",
  "license": "Apache-2.0",
  "publishConfig": {
    "access": "public"
  },
  "version": "0.0.0",
  "jest": {
    "testMatch": [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)"
    ],
    "clearMocks": true,
    "collectCoverage": true,
    "coverageReporters": [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text"
    ],
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": [
      "/node_modules/"
    ],
    "testPathIgnorePatterns": [
      "/node_modules/"
    ],
    "watchPathIgnorePatterns": [
      "/node_modules/"
    ],
    "reporters": [
      "default",
      [
        "jest-junit",
        {
          "outputDirectory": "test-reports"
        }
      ]
    ],
    "transform": {
      "^.+\\.[t]sx?$": [
        "ts-jest",
        {
          "tsconfig": "tsconfig.dev.json"
        }
      ]
    }
  },
  "types": "lib/index.d.ts",
  "stability": "stable",
  "jsii": {
    "outdir": "dist",
    "targets": {
      "python": {
        "distName": "cloudparing.aws-cdk-lib",
        "module": "cloudparing.aws_cdk_lib"
      }
    },
    "tsc": {
      "outDir": "lib",
      "rootDir": "src"
    }
  },
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"npx projen\"."
}
