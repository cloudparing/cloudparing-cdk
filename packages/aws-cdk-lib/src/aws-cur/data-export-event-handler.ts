import {
  BCMDataExportsClient,
  CreateExportCommand,
  CreateExportCommandInput,
  DeleteExportCommand,
  Export,
  ExportReference,
  ListExportsCommand,
  ListExportsResponse,
  UpdateExportCommand,
} from '@aws-sdk/client-bcm-data-exports';
import {
  CloudFormationCustomResourceCreateEvent,
  CloudFormationCustomResourceDeleteEvent,
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceHandler,
  CloudFormationCustomResourceUpdateEvent,
} from 'aws-lambda';
import {
  CompressionFormat,
  ExportVersioning,
} from './data-export-definition.types';
import { TimeUnit } from './report-definition.types';

export interface Cur2ExportParameters {
  readonly ServiceToken: string;
  readonly ExportType: 'COST_AND_USAGE';
  readonly ExportName: string;
  readonly ExportDescription: string;
  readonly S3Bucket: string;
  readonly S3Prefix: string;
  readonly S3Region: string;
  readonly TimeUnit: TimeUnit;
  readonly CompressionFormat: CompressionFormat;
  readonly ExportVersioning: ExportVersioning;
  readonly IncludeResourceIds: boolean;
  readonly SplitCostAllocationData: boolean;
  readonly SelectedColumns: string[];
}

export const handler: CloudFormationCustomResourceHandler = async (
  event: CloudFormationCustomResourceEvent,
): Promise<void> => {
  try {
    switch (event.RequestType) {
      case 'Create':
        return await createDataExport(
          event as CloudFormationCustomResourceCreateEvent,
        );
      case 'Update':
        return await updateDataExport(
          event as CloudFormationCustomResourceUpdateEvent,
        );
      case 'Delete':
        return await deleteDataExport(
          event as CloudFormationCustomResourceDeleteEvent,
        );
    }
  } catch (error) {
    // Handle error
    console.error('Error:', error);
    throw error;
  }
};
const createDataExport = async (
  event: CloudFormationCustomResourceCreateEvent,
): Promise<void> => {
  const props = event.ResourceProperties as Cur2ExportParameters;
  const client = new BCMDataExportsClient();
  const command = new CreateExportCommand({
    Export: cur2ExportParametersToExport(props),
  } as CreateExportCommandInput);
  await client.send(command);
};

const updateDataExport = async (
  event: CloudFormationCustomResourceUpdateEvent,
): Promise<void> => {
  const props = event.ResourceProperties as Cur2ExportParameters;
  const client = new BCMDataExportsClient();
  const foundExport = await findExportByName(client, props.ExportName);
  if (!foundExport) {
    throw new Error(`Export with name '${props.ExportName}' not found`);
  }

  const command = new UpdateExportCommand({
    ExportArn: foundExport.ExportArn,
    Export: cur2ExportParametersToExport(props),
  });
  await client.send(command);
};

const deleteDataExport = async (
  event: CloudFormationCustomResourceDeleteEvent,
): Promise<void> => {
  const props = event.ResourceProperties as Cur2ExportParameters;
  const client = new BCMDataExportsClient();
  const foundExport = await findExportByName(client, props.ExportName);
  if (!foundExport) {
    throw new Error(`Export with name '${props.ExportName}' not found`);
  }
  const command = new DeleteExportCommand({
    ExportArn: foundExport.ExportArn,
  });
  await client.send(command);
};

const cur2ExportParametersToExport = (props: Cur2ExportParameters): Export => {
  return {
    Name: props.ExportName,
    Description: props.ExportDescription || '',
    DataQuery: {
      QueryStatement: buildSQL(props.SelectedColumns),
      TableConfigurations: {
        COST_AND_USAGE_REPORT: {
          TIME_GRANULARITY: props.TimeUnit || 'DAILY',
          INCLUDE_RESOURCES: props.IncludeResourceIds ? 'TRUE' : 'FALSE',
          INCLUDE_SPLIT_COST_ALLOCATION_DATA: props.SplitCostAllocationData
            ? 'TRUE'
            : 'FALSE',
          INCLUDE_MANUAL_DISCOUNT_COMPATIBILITY: 'FALSE',
        },
      },
    },
    DestinationConfigurations: {
      S3Destination: {
        S3Bucket: props.S3Bucket,
        S3Prefix: props.S3Prefix,
        S3Region: props.S3Region,
        S3OutputConfigurations: {
          OutputType: 'CUSTOM',
          Format:
            props.CompressionFormat === 'Parquet' ? 'PARQUET' : 'TEXT_OR_CSV',
          Compression:
            props.CompressionFormat === 'Parquet' ? 'PARQUET' : 'GZIP',
          Overwrite:
            props.ExportVersioning === 'OVERWRITE_EXPORT'
              ? 'OVERWRITE_REPORT'
              : 'CREATE_NEW_REPORT',
        },
      },
    },
    RefreshCadence: {
      Frequency: 'SYNCHRONOUS',
    },
  };
};

const findExportByName = async (
  client: BCMDataExportsClient,
  name: string,
): Promise<ExportReference | undefined> => {
  let nextToken: string | undefined = undefined;

  do {
    const exports = (await client.send(
      new ListExportsCommand({
        MaxResults: 50,
        NextToken: nextToken,
      }),
    )) as ListExportsResponse;
    if (exports.Exports === undefined) {
      return undefined;
    }
    for (const exp of exports.Exports) {
      if (exp.ExportName === name) {
        return exp;
      }
    }
    nextToken = exports.NextToken;
  } while (nextToken);

  return undefined;
};

const buildSQL = (columns: string[]): string => {
  return `SELECT ${columns.join(', ')} FROM COST_AND_USAGE_REPORT`;
};
