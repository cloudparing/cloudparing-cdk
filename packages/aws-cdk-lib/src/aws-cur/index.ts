export * from './types';
export * from './report-definition';
export * from './stack';