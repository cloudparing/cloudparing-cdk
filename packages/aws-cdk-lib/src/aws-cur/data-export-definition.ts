import * as cdk from 'aws-cdk-lib';
import { CustomResource, Duration } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Architecture } from 'aws-cdk-lib/aws-lambda';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Provider } from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import {
  Cur2ExportProps,
  DataExportDefinitionProps,
} from './data-export-definition.types';
import { DataExportFunction } from './data-export-function';

const curRegion = 'us-east-1'; // cur only available in us-east-1
export class DataExportDefinition extends Construct {
  constructor(scope: Construct, id: string, props: DataExportDefinitionProps) {
    super(scope, id);

    // Add bucket policy https://docs.aws.amazon.com/cur/latest/userguide/cur-s3.html
    const policy = this.addBucketPolicy(props.bucket);
    if (policy) {
      this.node.addDependency(policy);
    }

    const dataExportFunction = new DataExportFunction(
      this,
      'DataExportFunction',
      {
        timeout: Duration.seconds(30),
        architecture: Architecture.ARM_64,
        memorySize: 1024,
        role: new iam.Role(this, 'DataExportFunctionRole', {
          assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
          inlinePolicies: {
            BcmDataExportPolicy: new iam.PolicyDocument({
              statements: [
                new iam.PolicyStatement({
                  sid: 'CreateCurExportsInDataExports',
                  effect: iam.Effect.ALLOW,
                  actions: ['bcm-data-exports:*'],
                  resources: ['*'],
                }),
                new iam.PolicyStatement({
                  sid: 'CurDataAccess',
                  effect: iam.Effect.ALLOW,
                  actions: ['cur:PutReportDefinition'],
                  resources: ['*'],
                }),
              ],
            }),
          },
        }),
      },
    );

    const customResourceProvider = new Provider(
      this,
      'DataExportDefinitionCustomResourceProvider',
      {
        onEventHandler: dataExportFunction,
        logRetention: RetentionDays.ONE_DAY,
      },
    );

    new CustomResource(this, 'DataExportDefinitionCustomResource', {
      serviceToken: customResourceProvider.serviceToken,
      resourceType: 'Custom::DataExportDefinition',
      properties: props.properties,
    });
  }

  private addBucketPolicy(
    bucket: cdk.aws_s3.IBucket,
  ): cdk.aws_s3.BucketPolicy | undefined {
    bucket.addToResourcePolicy(
      new cdk.aws_iam.PolicyStatement({
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: ['s3:PutObject', 's3:GetBucketPolicy'],
        principals: [
          new cdk.aws_iam.ServicePrincipal('billingreports.amazonaws.com'),
          new cdk.aws_iam.ServicePrincipal('bcm-data-exports.amazonaws.com'),
        ],
        resources: [bucket.bucketArn, bucket.arnForObjects('*')],
        conditions: {
          StringLike: {
            'aws:SourceArn': [
              `arn:${cdk.Aws.PARTITION}:cur:${curRegion}:${cdk.Aws.ACCOUNT_ID}:definition/*`,
              `arn:${cdk.Aws.PARTITION}:bcm-data-exports:${curRegion}:${cdk.Aws.ACCOUNT_ID}:export/*`,
            ],
            'aws:SourceAccount': `${cdk.Aws.ACCOUNT_ID}`,
          },
        },
      }),
    );

    return bucket.policy;
  }
}
export class Cur2ExportDefinition extends DataExportDefinition {
  constructor(scope: Construct, id: string, props: Cur2ExportProps) {
    super(scope, id, {
      ...props,
      properties: {
        ExportType: 'COST_AND_USAGE',
        ExportName: props.name,
        ExportDescription: props.description || '',
        S3Bucket: props.bucket.bucketName,
        S3Prefix: props.s3Prefix || '',
        S3Region: curRegion,
        TimeUnit: props.timeUnit || 'DAILY',
        CompressionFormat: props.compressionFormat || 'GZIP_CSV',
        ExportVersioning: props.exportVersioning || 'OVERWRITE_EXPORT',
        IncludeResourceIds: props.includeResourceIds || false,
        SplitCostAllocationData: props.splitCostAllocationData || false,
        SelectedColumns: props.columns || ALL_CUR_COLUMNS,
      },
    });
  }
}

export const ALL_CUR_COLUMNS = [
  'bill_bill_type',
  'bill_billing_entity',
  'bill_billing_period_end_date',
  'bill_billing_period_start_date',
  'bill_invoice_id',
  'bill_invoicing_entity',
  'bill_payer_account_id',
  'bill_payer_account_name',
  'cost_category',
  'discount',
  'discount_bundled_discount',
  'discount_total_discount',
  'identity_line_item_id',
  'identity_time_interval',
  'line_item_availability_zone',
  'line_item_blended_cost',
  'line_item_blended_rate',
  'line_item_currency_code',
  'line_item_legal_entity',
  'line_item_line_item_description',
  'line_item_line_item_type',
  'line_item_net_unblended_cost',
  'line_item_net_unblended_rate',
  'line_item_normalization_factor',
  'line_item_normalized_usage_amount',
  'line_item_operation',
  'line_item_product_code',
  'line_item_tax_type',
  'line_item_unblended_cost',
  'line_item_unblended_rate',
  'line_item_usage_account_id',
  'line_item_usage_account_name',
  'line_item_usage_amount',
  'line_item_usage_end_date',
  'line_item_usage_start_date',
  'line_item_usage_type',
  'pricing_currency',
  'pricing_lease_contract_length',
  'pricing_offering_class',
  'pricing_public_on_demand_cost',
  'pricing_public_on_demand_rate',
  'pricing_purchase_option',
  'pricing_rate_code',
  'pricing_rate_id',
  'pricing_term',
  'pricing_unit',
  'product',
  'product_comment',
  'product_fee_code',
  'product_fee_description',
  'product_from_location',
  'product_from_location_type',
  'product_from_region_code',
  'product_instance_family',
  'product_instance_type',
  'product_instancesku',
  'product_location',
  'product_location_type',
  'product_operation',
  'product_pricing_unit',
  'product_product_family',
  'product_region_code',
  'product_servicecode',
  'product_sku',
  'product_to_location',
  'product_to_location_type',
  'product_to_region_code',
  'product_usagetype',
  'reservation_amortized_upfront_cost_for_usage',
  'reservation_amortized_upfront_fee_for_billing_period',
  'reservation_availability_zone',
  'reservation_effective_cost',
  'reservation_end_time',
  'reservation_modification_status',
  'reservation_net_amortized_upfront_cost_for_usage',
  'reservation_net_amortized_upfront_fee_for_billing_period',
  'reservation_net_effective_cost',
  'reservation_net_recurring_fee_for_usage',
  'reservation_net_unused_amortized_upfront_fee_for_billing_period',
  'reservation_net_unused_recurring_fee',
  'reservation_net_upfront_value',
  'reservation_normalized_units_per_reservation',
  'reservation_number_of_reservations',
  'reservation_recurring_fee_for_usage',
  'reservation_reservation_a_r_n',
  'reservation_start_time',
  'reservation_subscription_id',
  'reservation_total_reserved_normalized_units',
  'reservation_total_reserved_units',
  'reservation_units_per_reservation',
  'reservation_unused_amortized_upfront_fee_for_billing_period',
  'reservation_unused_normalized_unit_quantity',
  'reservation_unused_quantity',
  'reservation_unused_recurring_fee',
  'reservation_upfront_value',
  'resource_tags',
  'savings_plan_amortized_upfront_commitment_for_billing_period',
  'savings_plan_end_time',
  'savings_plan_instance_type_family',
  'savings_plan_net_amortized_upfront_commitment_for_billing_period',
  'savings_plan_net_recurring_commitment_for_billing_period',
  'savings_plan_net_savings_plan_effective_cost',
  'savings_plan_offering_type',
  'savings_plan_payment_option',
  'savings_plan_purchase_term',
  'savings_plan_recurring_commitment_for_billing_period',
  'savings_plan_region',
  'savings_plan_savings_plan_a_r_n',
  'savings_plan_savings_plan_effective_cost',
  'savings_plan_savings_plan_rate',
  'savings_plan_start_time',
  'savings_plan_total_commitment_to_date',
  'savings_plan_used_commitment',
];
