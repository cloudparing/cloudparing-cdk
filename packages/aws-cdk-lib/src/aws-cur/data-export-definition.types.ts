import * as cdk from 'aws-cdk-lib';

export type ExportTimeUnit = 'HOURLY' | 'DAILY' | 'MONTHLY';

export type CompressionFormat = 'GZIP_CSV' | 'Parquet';
export type ExportVersioning = 'CREATE_NEW_EXPORT' | 'OVERWRITE_EXPORT';

export interface DataExportDefinitionProps
  extends Cur2ExportOptions,
  DataExportStorageOptions {
  /**
   * Properties to pass to the Lambda
   *
   * @default - No properties.
   */
  readonly properties?: {
    [key: string]: any;
  };
}

export interface Cur2ExportOptions {
  /**
   * The name of this specific data export.
   */
  readonly name: string | undefined;

  /**
   * The description for this specific data export.
   */
  readonly description?: string;
}

export interface Cur2ContentOptions {
  /**
   * Include a column containing the unique AWS resource ID for applicable
   * line items. Including individual resource IDs in your export might
   * increase the file size
   *
   * @default false
   */
  readonly includeResourceIds?: boolean;

  /**
   * Include detailed cost and usage for shared resources for cost allocation
   * (only available for Amazon ECS). Including these resources introduces new
   * rows and columns in the Cost and Usage Report and might increase the file size
   *
   * @default false
   */
  readonly splitCostAllocationData?: boolean;

  /**
   * The time granularity for how you want the line items in the export to be aggregated.
   *
   * @default HOURLY
   *
   */
  readonly timeUnit?: ExportTimeUnit;

  /**
   * The columns that you want to include in your data query. This allows
   * you to create exports with consistent schemas, remove sensitive cost data,
   * and reduce the file size of the export. By default, all columns are selected
   *
   * @default false
   */
  readonly columns?: string[];
}

export interface Cur2DeliveryOptions {
  /**
   * The compression type and file format for your export.
   *
   * @default GZIP_CSV
   *
   */
  readonly compressionFormat?: CompressionFormat;

  /**
   * Whether you want each version of the data export file to overwrite the
   * previous version or to be delivered in addition to the previous versions.
   *
   * @default OVERWRITE_EXPORT
   */
  readonly exportVersioning?: ExportVersioning;
}

export interface DataExportStorageOptions {
  /**
   * The S3 bucket where your data export will be stored.
   *    Length Constraints: Maximum length of 256.
   *    Pattern: [A-Za-z0-9_\.\-]+
   *
   */
  readonly bucket: cdk.aws_s3.IBucket;

  /**
   * The prefix that AWS adds to the report name when AWS delivers the report. Your prefix can't include spaces.
   *    Length Constraints: Maximum length of 256.
   *    Pattern: [A-Za-z0-9_\.\-]*
   *
   * @default ""
   *
   */
  readonly s3Prefix?: string;
}

export interface Cur2ExportProps
  extends Cur2ExportOptions,
  Cur2ContentOptions,
  Cur2DeliveryOptions,
  DataExportStorageOptions {}
