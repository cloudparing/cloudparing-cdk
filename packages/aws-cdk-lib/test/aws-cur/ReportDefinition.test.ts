import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Bucket } from 'aws-cdk-lib/aws-s3';

import { ReportDefinition } from '../../lib/aws-cur';

test('snapshot: cur report defaults', () => {
  const stack = new Stack();

  const curBucket = new Bucket(stack, 'CurBucketId', {
    bucketName: 'cur-bucket-name',
  });
  new ReportDefinition(stack, 'ReportDefinitionId', {
    reportName: 'ReportDefinitionName',
    s3Bucket: curBucket,
    s3Prefix: 'ReportDefinitionPrefix',
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});


test('snapshot: cur report spec\'d out', () => {
  const stack = new Stack();

  const curBucket = new Bucket(stack, 'CurBucketId', {
    bucketName: 'cur-bucket-name',
  });
  new ReportDefinition(stack, 'ReportDefinitionId', {
    reportName: 'ReportDefinitionName',
    s3Bucket: curBucket,
    s3Prefix: 'ReportDefinitionPrefix',
    timeUnit: 'MONTHLY',
    additionalArtifacts: ['REDSHIFT'],
    additionalSchemaElements: ['ReportDefinitionAdditionalSchemaElements'],
    billingViewArn: 'ReportDefinitionBillingViewArn',
    format: 'Parquet',
    refreshClosedReports: false,
    reportVersioning: 'CREATE_NEW_REPORT',
    compression: 'ZIP',
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});

