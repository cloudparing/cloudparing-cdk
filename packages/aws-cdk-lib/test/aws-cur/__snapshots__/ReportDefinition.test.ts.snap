// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: cur report defaults 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CurBucketId63DADEFC": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "cur-bucket-name",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "CurBucketIdPolicy098FAE91": {
      "Properties": {
        "Bucket": {
          "Ref": "CurBucketId63DADEFC",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetBucketAcl",
                "s3:GetBucketPolicy",
              ],
              "Condition": {
                "StringEquals": {
                  "aws:SourceAccount": {
                    "Ref": "AWS::AccountId",
                  },
                  "aws:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":cur:us-east-1:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":definition/*",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "billingreports.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "CurBucketId63DADEFC",
                  "Arn",
                ],
              },
            },
            {
              "Action": "s3:PutObject",
              "Condition": {
                "StringEquals": {
                  "aws:SourceAccount": {
                    "Ref": "AWS::AccountId",
                  },
                  "aws:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":cur:us-east-1:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":definition/*",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "billingreports.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "CurBucketId63DADEFC",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "ReportDefinitionIdReportDefinitionIdResource5880FDCB": {
      "DependsOn": [
        "CurBucketIdPolicy098FAE91",
      ],
      "Properties": {
        "Compression": "GZIP",
        "Format": "textORcsv",
        "RefreshClosedReports": true,
        "ReportName": "ReportDefinitionName",
        "ReportVersioning": "OVERWRITE_REPORT",
        "S3Bucket": {
          "Ref": "CurBucketId63DADEFC",
        },
        "S3Prefix": "ReportDefinitionPrefix",
        "S3Region": "\${Token[AWS",
        "TimeUnit": "DAILY",
      },
      "Type": "AWS::CUR::ReportDefinition",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`snapshot: cur report spec'd out 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CurBucketId63DADEFC": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "cur-bucket-name",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "CurBucketIdPolicy098FAE91": {
      "Properties": {
        "Bucket": {
          "Ref": "CurBucketId63DADEFC",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetBucketAcl",
                "s3:GetBucketPolicy",
              ],
              "Condition": {
                "StringEquals": {
                  "aws:SourceAccount": {
                    "Ref": "AWS::AccountId",
                  },
                  "aws:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":cur:us-east-1:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":definition/*",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "billingreports.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "CurBucketId63DADEFC",
                  "Arn",
                ],
              },
            },
            {
              "Action": "s3:PutObject",
              "Condition": {
                "StringEquals": {
                  "aws:SourceAccount": {
                    "Ref": "AWS::AccountId",
                  },
                  "aws:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":cur:us-east-1:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":definition/*",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "billingreports.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "CurBucketId63DADEFC",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "ReportDefinitionIdReportDefinitionIdResource5880FDCB": {
      "DependsOn": [
        "CurBucketIdPolicy098FAE91",
      ],
      "Properties": {
        "AdditionalArtifacts": [
          "REDSHIFT",
        ],
        "AdditionalSchemaElements": [
          "ReportDefinitionAdditionalSchemaElements",
        ],
        "BillingViewArn": "ReportDefinitionBillingViewArn",
        "Compression": "ZIP",
        "Format": "Parquet",
        "RefreshClosedReports": true,
        "ReportName": "ReportDefinitionName",
        "ReportVersioning": "CREATE_NEW_REPORT",
        "S3Bucket": {
          "Ref": "CurBucketId63DADEFC",
        },
        "S3Prefix": "ReportDefinitionPrefix",
        "S3Region": "\${Token[AWS",
        "TimeUnit": "MONTHLY",
      },
      "Type": "AWS::CUR::ReportDefinition",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
